Make Superior HttpHandler

2lazy2late opened the pull request:
Prepare for different services on different routes

Commit:
Make Superior HttpHandler

Prepare for different services on different routes

Commit:
Just implement functions quickly

Never mind exceptions or anything else

johngrib commented on app/src/main/java/com/codesoom/assignment/SuperHttpHandler.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public void handle(HttpExchange exchange) throws IOException {
+        // 1. Method - GET, POST, PUT/PATCH, DELETE, ...
+        // 2. Path - "/", "/tasks", "/tasks/1", ...
+        // 3. Headers, Body(Content)
```
구현을 위한 메모였다면 깔끔하게 삭제해 주도록 합시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/SuperHttpHandler.java:
```diff
@@ -0,0 +1,38 @@
+                .lines()
+                .collect(Collectors.joining("\n"));
+
+        if(path.startsWith("/tasks")){
+            TaskHandler taskHandler = new TaskHandler();
+            taskHandler.handler(method, path, body, exchange);
+        }else {
```
사용하고 계신 IDE에서 포매터를 사용하는 방법을 조사해 보세요. 단축키로 실행할 수도 있고, 파일을 저장할 때마다 자동으로 돌아가게 만들 수도 있습니다. 조사를 완료하면 그 때부터는 이런 종류의 스페이스 문제는 신경쓰지 않게 될 거에요.

```suggestion
        if (path.startsWith("/tasks")) {
            TaskHandler taskHandler = new TaskHandler();
            taskHandler.handler(method, path, body, exchange);
        } else {
```

2lazy2late replied:
[google-java-format 사용법](https://johngrib.github.io/wiki/google-java-format/) 을 참조하여
save action 플러그인과 조합하여 적용하였습니다.

작동이 잘 되는 줄 알았으나..
일부 파일 ex) App.java 에만 formatter가 작동을 하고
다른 파일에는 강제로 formatter를 실행시켜도 작동하지 않습니다.

다른 방법을 좀 더 찾아보겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/SuperHttpHandler.java:
```diff
@@ -0,0 +1,38 @@
+            TaskHandler taskHandler = new TaskHandler();
+            taskHandler.handler(method, path, body, exchange);
+        }else {
+            // If response length has the value -1 then no response body is being sent.
```
코드숨을 진행하는 동안에는 주석과 커밋 메시지는 영어가 아니라 한국어로 작성해 보세요. 지금은 영어공부가 중요하지 않습니다. 한국어로 먼저 명확하고 분명하게 작성하는 연습을 하고, 충분히 좋은 문장을 작성할 수 있게 되면 그 때부터 영어로 작성해도 늦지 않습니다.

2lazy2late replied:
주석과 커밋 메시지를 한글로 작성하도록 하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskHandler.java:
```diff
@@ -0,0 +1,130 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.sun.net.httpserver.HttpExchange;
+
+import java.io.*;
```
`import *`의 사용을 지양하고, 어떤 것을 `import` 하는지 분명히 명시해 주도록 합시다. 보통은 이게 자동으로 될 텐데, 이걸 막기위해 IDE에서 설정을 할 수 있을 거에요.

2lazy2late replied:
save action 플러그인의 import files optimize 를 사용하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskHandler.java:
```diff
@@ -0,0 +1,130 @@
+
+            // Specify a value if task id exists in the path, and initializes to 0 if not.
+            long lTaskId = 0L;
+            lTaskId = arrPath.length == 3 ? Long.parseLong(arrPath[2]) : 0;
```
이 라인에 매직 넘버가 3개 있습니다. `3`, `2`, `0` 이 숫자들의 의미는 무엇인가요? 코드를 읽는 사람이 이 숫자들의 의미를 파악하기 위해서 읽어야 하는 다른 곳의 코드는 몇 줄이 있을까요? 이걸 한줄만 읽고도, 또는 위아래 3줄 정도만 읽고도 이해하게 하기 위해서는 어떻게 하면 좋을까요?

2lazy2late replied:
```
  private static final int LENGTH_OF_PATH_WITH_VALUE = 3;
  private static final int LOCATION_OF_VALUE_IN_PATH = 2;
  private static final long DEFAULT_VALUE_ASSIGNED_TO_LONG_TYPE = 0L;


    // 경로에 taskId 값이 존재한다면 해당 값으로, 존재하지 않는다면 long 타입의 기본값인 0L으로 선언한다.
    return arrPath.length == LENGTH_OF_PATH_WITH_VALUE
        ? Long.parseLong(arrPath[LOCATION_OF_VALUE_IN_PATH])
        : DEFAULT_VALUE_ASSIGNED_TO_LONG_TYPE;
```



해당 라인만 보고도 명확히 이해할 수 있도록 상수명을 변경하고 한글 주석을 추가하였습니다.

johngrib replied:
좋습니다. 조금 더 좋아졌군요. 그런데 지금도 좀 아쉬운 점이 있어요. `LENGTH_OF_PATH_WITH_VALUE` 는 왜 3이어야 하는가? 같은 거죠. 왜 3이어야 하나요?

2lazy2late replied:
이미 Request Validation 에서 Path 패턴 체크를 했으니
굳이 LENGTH_OF_PATH_WITH_VALUE 을 사용할 이유가 없어졌네요..
해당 상수를 제거한 로직으로 변경하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskHandler.java:
```diff
@@ -0,0 +1,130 @@
+
+            String responseContent = null;
+
+            // Method
+            switch (method){
```
바로 아래에 `switch (method)`가 있기 때문에 `// Method`는 불필요해 보입니다.

2lazy2late replied:
삭제하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskHandler.java:
```diff
@@ -0,0 +1,130 @@
+                    responseContent = getTasks(lTaskId);
+                    break;
+                case "POST" :
+                    index++;
```
여기에서 인덱스를 증가시키는 이유는 무엇인가요? 이 인덱스는 무엇을 뜻하나요?

2lazy2late replied:
새로운 taskId 를 생성하는데 필요한 index 값이었는데
코드만 보고서는 알기 쉽지 않아서
class 로 만들어서 method 명을 명시적으로 선언하였습니다.

```
public class TaskIndex {
  private static Long lastIndex;

  public TaskIndex() {
    lastIndex = 0L;
  }

  public Long getLastIndex() {
    return lastIndex;
  }

  public Long getNextIndex() {
    return lastIndex + 1;
  }

  public void addLastIndex() {
    lastIndex += lastIndex + 1;
  }
}
```

johngrib replied:
잘 만드신 것 같습니다. 그런데 코멘트 내부의 코드를 읽어보니 `TaskIndex`에서 `Task`가 없어도 괜찮지 않을까요? 이 클래스의 진정한 역할은 무엇일까요?

2lazy2late replied:
역할 자체로만 본다면 TASK에만 국한된 기능들이 아니기에 class 명에 task가 빠져도 될 것 같습니다.
실질적인 용도는 index 관리인데 다시 고민해서 명명해 보도록 하겠습니다.

johngrib replied:
이 프로젝트 내부를 보고 있는 관점에서 한 걸음 물러나서 이 클래스를 별도의 라이브러리라 생각해 보세요. 아마 좀 더 다른 생각이 떠오를지도 모릅니다. 지금 패키지 이름도 `assignment`인데 이 패키지 이름도 바꾸거나 분리하거나 할 수 있겠죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskHandler.java:
```diff
@@ -0,0 +1,130 @@
+    private String delTask(Long taskId) throws Exception {
+        if(!taskMap.containsKey(taskId)){
+            throw new Exception("This key is not exist");
+        } else{
```
이 메소드는 `else`가 없다면 좀 더 깔끔해 보일 수 있을 것 같군요. 실제로 `else`가 없어도 잘 작동할 것입니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/TaskHandler.java:
```diff
@@ -0,0 +1,130 @@
+
+    private String delTask(Long taskId) throws Exception {
+        if(!taskMap.containsKey(taskId)){
+            throw new Exception("This key is not exist");
```
코드 이전에 문제 상황을 곰곰이 생각해 봅시다. 키가 없는 것이 예외적인 상황일까요? 우리는 매일 구글에서 온갖 것을 검색하죠. 그런데 검색 결과가 없을 때 어떻게 되나요? 예외를 던지나요? 아니면 그냥 검색 결과가 없다고 하나요?

만약 예외를 던져야 하는 상황이 맞다면 어째서 그런 것인지를 자세히 적어 보세요. 저에게 보여주려고가 아니라 의사결정을 점검한다고 생각하고 잘 작성해 보세요.

2lazy2late replied:
예외를 던지지(throw) 않고 catch 해서
적절한 response status와 함께 메시지를 리턴할 수 있게끔 하였습니다.

johngrib replied:
왜 그렇게 하셨나요? 행동은 설명하셨는데 어떤 생각을 해서 그러한 결론에 도달했는지도 이야기해주시면 좋을 거에요. 자신의 코드와 관련된 의사결정을 설명하는 것도 연습해보도록 합시다.

johngrib reviewed (commented):
반갑습니다!

Commit:
구조를 변경하라

기존 com.codesoom.assignment 패키지 아래에 todo 패키지를 만들어서
그 하위에 관련된 code 들을 위치시켰다.
또한 controller, service 를 생성하여 mvc 패턴을 명확히 하였다.
전체 task 의 갯수를 관리하지 않고 마지막 생성된 task id를 관리하기 위하여
TaskIndex class 를 생성하여 사용하였다.

Commit:
에러 컨트롤과 응답방식 변경에 대비하라

컨트롤러에서 바로 응답을 보내기 위해 Response 클래스를 작성해 두었음.

johngrib reviewed (commented):
3032b47b22d44f6b008b9c36ff9efb46de84740d 커밋이 포함하고 있는 변경 내용을 잘 읽어보면 커밋의 주제와 맞지 않는 내용도 다소 포함되어 있습니다. 커밋 메시지 본문은 열심히 작성하셨지만 `구조를 변경하라`는 커밋 타이틀은 추상적이어서 어떤 구조를 어떻게 변경했는지에 대해 파악하려면 역시 내용을 열심히 읽어보는 수 밖에 없습니다. 커밋 타이틀에 좀 더 구체적인 내용을 포함시켜 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/models/Response.java:
```diff
@@ -0,0 +1,26 @@
+import java.io.OutputStream;
+import java.util.Objects;
+
+public class Response {
```
어떤 응답인지 클래스 이름으로 더 구체적으로 표현할 수 있으면 좋겠군요.

2lazy2late replied:
해당 클래스는.. 실제로 응답용인데
HttpResponse ? 정도로 바꿔야 할까요..
조금 더 힌트를 부탁드립니다

johngrib replied:
제가 답을 알려드리는 것이 아닙니다. 그러면 제가 이야기하는대로 코딩하게 되겠죠! 클래스 이름의 대안을 3가지 정도 떠올려보시고, 각각의 장점과 단점에 대해 설명해보세요. 그리고 그 중 하나를 선택하고 왜 선택했는지도 써보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/models/Response.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.Objects;
+
+public class Response {
+  private final int iResponseStatusCode;
+  private final String sResponseContent;
```
헝가리안 변수 명명법은 가능한 한 사용하지 않도록 합시다. 좋지 않은 습관입니다. 변수 prefix에 i, s 를 넣지 않아도 타입이 구별되므로 이렇게 작성하지 않아도 괜찮습니다.

2lazy2late replied:
다음 커밋 때 변경하도록 하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/SuperHttpHandler.java:
```diff
@@ -0,0 +1,39 @@
+    } else {
+      // If response length has the value -1 then no response body is being sent.
+      // response 길이가 -1 이라면 response body 가 전송되지 않습니다.
+      exchange.sendResponseHeaders(404, -1);
```
`-1` 같은 센티넬 밸류를 사용하는 것보다 더 나은 방법을 궁리해 보도록 합시다. 더 나은 방법이 있을 거에요.

2lazy2late replied:
변경하였습니다.

johngrib replied:
어떤 방법을 썼나요? 그냥 변경했다고 작성하셨는데요, 무엇을 어떻게 왜 어떤 방식으로 변경했는지 설명해 보세요. 의사결정과정을 다른 사람에게 보여주는 훈련을 해 봅시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/TaskHandler.java:
```diff
@@ -0,0 +1,49 @@
+            outputStream.write(responseContent.getBytes());
+            outputStream.flush();
+            outputStream.close();
+        } catch(NumberFormatException e){
+            e.printStackTrace();
+        } catch (IOException ex) {
+            throw new RuntimeException(ex);
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw new RuntimeException(e);
+        }
```
여기에서 가능한 한 catch 문을 사용하지 않는 방향으로 고쳐야 한다고 생각해 봅시다. 어떻게 해야 좋을까요?

2lazy2late replied:
컨트롤러에서 미리 예외를 응답으로 리턴하도록 처리하여서 없애보았습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/controllers/TaskController.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.io.IOException;
+
+public class TaskController {
```
TaskController의 존재의의는 무엇인가요?

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/controllers/TaskController.java:
```diff
@@ -0,0 +1,36 @@
+    return taskService.addTask(task);
+  }
+
+  public String modTask(String sRequestPath, String sRequestBody) throws Exception {
```
메소드 이름을 축약하는 것은 좋지 않습니다. 물론 너무 길어지면 줄일 방법을 고민해야겠지만 이건 너무 짧군요. 가능한 한 축약 없이 메소드 이름을 복원해 봅시다.

2lazy2late replied:
명시적으로 변경하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/models/TaskIndex.java:
```diff
@@ -0,0 +1,21 @@
+    return lastIndex + 1;
+  }
+
+  public void addLastIndex() {
+    lastIndex += lastIndex + 1;
+  }
```
여기에서는 스레드 안전성을 고민해 볼 필요가 있어 보입니다.

2lazy2late replied:
synchronized 로 인스턴스 동기화 시켜서 해결하였습니다.

Commit:
RequestValidation Class 완성

모든 Request에 대한 정합성 체크로직을 작성하였다

Commit:
Response 클래스의 적용

RequestValidation 클래스 작성 후 미리 만들어둔 Response 클래스로 응답 보내기

Commit:
예외에 대응하라

예외코드와 메시지를 새로 정의하고 발생시키는 로직을 구체화하였다.

johngrib reviewed (commented):
코드숨 과정을 진행하는 동안에는 이 버튼은 직접 클릭하지 마시고 제가 클릭하게 해 주세요. 한번에 여러 PR을 리뷰하다보면 꽤 헷갈리거든요.

![image](https://user-images.githubusercontent.com/1855714/173098255-7dc41f07-10ac-4bfb-b1fc-3f18c6229d15.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/TaskHandler.java:
```diff
@@ -15,11 +16,16 @@ public class TaskHandler {
     TaskController taskController = new TaskController(taskService);
 
-    public void handler(String sRequestMethod, String sRequestPath, String sRequestBody, HttpExchange exchange) {
+    public void handler(String sRequestMethod, String sRequestPath, String sRequestBody, String sRequestQuery, HttpExchange exchange) {
```
입력 파라미터가 꽤 많아서 읽기 어렵게 느껴집니다. 파라미터의 수를 1개나 2개로 줄여야 한다는 서브 과제가 주어진다면 어떻게 하면 좋을까요?

2lazy2late replied:
말씀하신 고민을 하면서 작성했었는데
trade off가 좀 있는 것 같아서 파라미터를 많이 받았습니다.
파라미터를 많이 받는 대신 중복된 코드를 없앴다고 생각했는데
무엇이 더 나은지, 다른 대안이 있는지를 고민해보겠습니다.

johngrib replied:
좋습니다! 그 고민을 혼자서 떠올렸다가 그대로 코드로 옮기지 마시고, 글로 작성해 보세요. 고민한 과정, 결론을 낸 과정을 전부 표현해보는 것이 필요할 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/models/RequestValidation.java:
```diff
@@ -0,0 +1,109 @@
+    boolean isRequestQueryExist = !isItHasNoContent(sRequestQuery);
+
+    // query 존재 허용하지 않음
+    if(isRequestQueryExist) return new RequestValidation(false, "URI 에 Query 는 허용되지 않습니다");
```
무슨 일이 있어도 if 에서 중괄호를 생략하지 마세요. 한 줄로 줄여서 얻는 이득이 생각보다 크지 않습니다.

2lazy2late replied:
주의하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/models/RequestValidation.java:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  public RequestValidation validationCheck(
+      String sRequestMethod, String sRequestPath, String sRequestBody, String sRequestQuery) {
```
헝가리안 변수 명명법을 사용하지 않는 것이 좋습니다. 파라미터 이름을 모두 수정해 주세요.

2lazy2late replied:
모두 수정하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/models/RequestValidation.java:
```diff
@@ -0,0 +1,109 @@
+      }
+    }
+
+    String pathValidMsg = isPathFormatCorrect ? "CORRECT" : "비 정상적인 경로입니다.";
```
이 코드 한 줄에 대해 설명이 듣고 싶습니다. `pathValidMsg`의 역할은 무엇인가요?

2lazy2late replied:
RequestValidation 클래스의 resultMsg에 할당됩니다
해당 메시지는 TaskHandler에서 리턴하는 Response에 들어가게 됩니다

johngrib replied:
`resultMsg`는 누가 읽게 되나요? 컴퓨터가 읽나요? 아니면 사람이 읽나요? 컴퓨터가 읽는다면 어디에 있는 어떤 역할의 컴퓨터가 읽게 될까요? `"CORRECT"`와 `"비 정상적인 경로입니다."`는 같은 독자를 갖는 메시지라 보기엔 꽤나 이질적으로 느껴집니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/models/RequestValidation.java:
```diff
@@ -0,0 +1,109 @@
+  }
+  public RequestValidation isItPossibleToConvertToTask(String content){
+
+    try{
+      Task changedTask;
+      changedTask = objectMapper.readValue(content, Task.class);
+
+      String parsedTaskId = String.valueOf(changedTask.getId());
+
+      if(isItHasNoContent(parsedTaskId) || "null".equals(parsedTaskId)){
+        return new RequestValidation(true, "contentBody 가 Task Class 타입으로 정상적으로 변환됨");
+      }else{
+        return new RequestValidation(false, "taskId 값을 지정하지 마십시오");
+      }
+    }catch (Exception e){
+      return new RequestValidation(false, "contentBody 는 Task Class 타입으로 변환할 수 없습니다");
```
`RequestValidation`의 두 번째 인자는 누가 읽는 메시지인가요?

2lazy2late replied:
해당 인자는 TaskHandler에서 리턴하는 Response에 들어가게 됩니다

johngrib replied:
앗! 질문에 대한 대답이 아닌 것 같아요.

Commit:
Request 객체를 생성하여 중복되는 코드를 줄여라

Request 객체를 생성하여 생성자로 하여금 자주 사용하는 변수를 선언하게 하였다.
코드 내에 존재하는 헝가리안 명명법으로 이뤄진 변수명을 모두 바꾸었다.

Commit:
헝가리안 명명법으로 작성한 코드를 변경하라

변수명에서 타입을 알아볼 수 있게 하는 명명법을 지양하고
역할만 표현하도록 하였다.

johngrib commented on .github/workflows/ci.yml:
```diff
@@ -1,6 +1,6 @@
 name: CI
 
-on: [push, pull_request]
+on: [ push, pull_request ]
```
이건 불필요한 수정 같아요.

johngrib commented on app/src/main/java/com/codesoom/assignment/SuperHttpHandler.java:
```diff
@@ -0,0 +1,32 @@
+      Response response = taskHandler.handler(request);
+      response.sendResponse(exchange);
+    } else {
+      // If response length has the value -1 then no response body is being sent.
+      // response 길이가 -1 이라면 response body 가 전송되지 않습니다.
```
이 주석은 불필요해 보입니다. 만약 필요하다면 무엇 때문일까요?

```suggestion
```

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/TaskHandler.java:
```diff
@@ -0,0 +1,48 @@
+public class TaskHandler {
+
+
+    TaskService taskService = new TaskService();
+    TaskController taskController = new TaskController(taskService);
```
```suggestion
    private final TaskService taskService = new TaskService();
    private final TaskController taskController = new TaskController(taskService);
```

johngrib commented on app/src/main/java/com/codesoom/assignment/todo/TaskHandler.java:
```diff
@@ -0,0 +1,48 @@
+
+        switch (requestMethod) {
+            case "GET" -> {
+                return taskController.getTasks(requestPath);
```
getTasks 가 String이 아니라 객체를 받게 한다면 어떨까요? 그렇게 작업한다면 그 객체는 어떤 구조를 가져야 좋을까요?
